lib/src/annotate.rs

```diff
@@ -0,0 +1,335 @@
+    }
+    let file_path: RepoPathBuf = file_path.unwrap();
+    let store = repo.store();
+
+    let current_commit = store.get_commit(starting_commit_id)?;
+    let current_tree = current_commit.tree()?;
+    let original_file_id = get_file_id(current_commit.id(), &current_tree, &file_path)?;
+    if original_file_id.is_none() {
+        return Err(AnnotateError::FileNotFound(file_name.to_string()));
+    }
+    let original_contents =
+        get_file_contents(repo.store(), &file_path, &original_file_id.unwrap())?;
+    let num_lines = original_contents.lines().count();
+    let mut partial_results = PartialResults::new(starting_commit_id, num_lines);
+
+    process_commits(repo, current_commit.id(), &file_path, &mut partial_results)?;
+
+    Ok(partial_results.convert_to_results(&original_contents))
+}
+
+fn process_commits(
+    repo: &ReadonlyRepo,
+    starting_commit_id: &CommitId,
+    file_name: &RepoPath,
+    results: &mut PartialResults,
+) -> Result<(), AnnotateError> {
+    let mut current_commit = Some(repo.store().get_commit(starting_commit_id)?);
+    while let Some(next_commit) = current_commit {
+        process_commit(&next_commit, repo, file_name, results)?;
+        let next_commit_id = results.get_next_commit();
+        match next_commit_id {
+            None => {
+                current_commit = None;
+            }
+            Some(next_id) => {
+                current_commit = Some(repo.store().get_commit(&next_id)?);
+            }
+        }
+    }
+    Ok(())
+}
+
+fn process_commit(
+    commit: &Commit,
+    repo: &ReadonlyRepo,
+    file_name: &RepoPath,
+    results: &mut PartialResults,
+) -> Result<(), AnnotateError> {
+    let current_tree = repo.store().get_commit(commit.id())?.tree()?;
+    let current_file_id = get_file_id(commit.id(), &current_tree, file_name)?.unwrap();
+
+    for parent in commit.parents() {
+        let parent_commit = parent?;
+        let parent_tree = parent_commit.tree()?;
+        let parent_file_id = get_file_id(parent_commit.id(), &parent_tree, file_name)?;
+
+        if let Some(pfid) = parent_file_id {
+            process_file_ids(
+                repo.store(),
+                results,
+                file_name,
+                &current_file_id,
+                commit.id(),
+                &pfid,
+                parent_commit.id(),
+            )?;
+        }
+    }
+    results.drain_remaining_for_commit_id(commit.id());
+
+    Ok(())
+}
+
+fn process_file_ids(
+    store: &Store,
+    results: &mut PartialResults,
+    file_name: &RepoPath,
+    current_file_id: &FileId,
+    current_commit_id: &CommitId,
+    parent_file_id: &FileId,
+    parent_commit_id: &CommitId,
+) -> BackendResult<()> {
+    if current_file_id == parent_file_id {
+        results.swap_full_commit_id(current_commit_id, parent_commit_id);
+        return Ok(());
+    }
+    let current_contents = get_file_contents(store, file_name, current_file_id)?;
```

Comment:
nit: might be better to cache original content in `local_line_map` instead of fetching again.

Reply:
do you mean cache the contents of each commit still in local_line_map?
Or do you actually mean the original content (content of the working copy)?

Reply:
Cache the content of each commit. I assume it will be consumed soon (as long as the graph is linear.) Cache of the original content might also be good.

Reply:
Added a file cache